src/util/RatelimitBucket.cpp

```diff
@@ -0,0 +1,45 @@
+#include "RatelimitBucket.hpp"
+
+#include <QTimer>
+
+namespace chatterino {
+
+RatelimitBucket::RatelimitBucket(int budget, int cooldown,
+                                 std::function<void(QString)> callback,
+                                 QObject *parent)
+    : QObject(parent)
+    , budget_(budget)
+    , cooldown_(cooldown)
+    , callback_(callback)
+{
+}
+
+void RatelimitBucket::send(QString message)
```

Comment:
q: Could a situation like this happen?
  1. `RatelimitBucket::send()` is called
  2. it appends the message/channel
  3. meanwhile `handleOne`'s empty check passed and it takes the element for itself
  4. then `send` calls `handleOne` with nothing in the queue

I'm scared that this could happen in extremely random cases and crash the app with a mysterious message that shouldn't happen.

Reply:
> I'm scared that this could happen in extremely random cases and crash the app with a mysterious message that shouldn't happen.

If anything, I'm pretty sure that a crash wouldn't happen. As stated by [Qt documentation](https://doc.qt.io/qt-5/qlist.html#removeFirst):

> The list must not be empty. If the list can be empty, call isEmpty() before calling this function.

Which is exactly what we're doing:

https://github.com/Chatterino/chatterino2/blob/658f28f31eb771ca7fcd156b524de5addb8c3072/src/util/RatelimitBucket.cpp#L29-L34